src/gbinder_servicemanager.c

```diff
@@ -433,6 +504,81 @@ gbinder_servicemanager_cancel(
     }
 }
 
+gulong
+gbinder_servicemanager_add_registration_handler(
+    GBinderServiceManager* self,
+    const char* name,
+    GBinderServiceManagerRegistrationFunc func,
+    void* user_data) /* Since 1.0.13 */
+{
+    gulong id = 0;
+
+    if (G_LIKELY(self) && G_LIKELY(func) && name && name[0]) {
+        GBinderServiceManagerClass* klass =
+            GBINDER_SERVICEMANAGER_GET_CLASS(self);
+        GBinderServiceManagerPriv* priv = self->priv;
+        GBinderServiceManagerWatch* watch = NULL;
+        char* sig;
+
+        watch = g_hash_table_lookup(priv->watch_table, name);
+        if (!watch) {
+            watch = gbinder_servicemanager_watch_new(name);
+            g_hash_table_insert(priv->watch_table, watch->name, watch);
+        }
+        if (!watch->watched) {
+            watch->watched = klass->watch(self, name);
+            if (watch->watched) {
+                GDEBUG("Watching %s", watch->name);
+            } else {
+                GWARN("Failed to watch %s", watch->name);
+            }
+        }
+
+        sig = g_strconcat(SIGNAL_REGISTRATION_NAME "::", watch->detail, NULL);
+        id = g_signal_connect(self, sig, G_CALLBACK(func), user_data);
+        g_free(sig);
```

Comment:
You could avoid the concatenation with:
`id = g_signal_connect_closure_by_id(gbinder_servicemanager_signals[SIGNAL_REGISTRATION], (watch->quark = g_quark_from_string(watch->detail)), g_cclosure_new(G_CALLBACK(func), user_data, NULL), FALSE);`

Reply:
I was considering different ways to optimize this part (e.g. a fixed-size buffer on stack since the lengths of all components are known at compile time) but decided that there's nothing to optimize here - this operation is likely to be performed just once in the lifetime of the process. So I opted for readability.